### PR TITLE
OCPBUGS-64688: use ObservedGeneration to determine Progressing status

### PR DESCRIPTION
**Pull request by sg00dwin**
## Summary
- Replace `IsAvailableAndUpdated` with `ObservedGeneration < Generation` check for the Progressing condition
- `IsAvailableAndUpdated` used `UpdatedReplicas == Replicas`, which falsely triggers during node reboots when pods are temporarily disrupted — even though the operator made no spec changes
- `ObservedGeneration < Generation` only fires when the operator has updated the deployment spec and the deployment controller hasn't processed it yet
- Extract `checkDeploymentGenerationProgress` helper for testability and add unit tests

## Root Cause
During node reboots (MCO draining nodes), console pods are terminated and replacements are created. During the ~30-second gap, `UpdatedReplicas != Replicas`, causing `IsAvailableAndUpdated` to return false and the operator to report `Progressing=True`. The operator made no changes — the disruption is external.

## Why This Fix
`ObservedGeneration < Generation` precisely detects operator-initiated spec changes:
- **Node reboot (no spec change):** Generation unchanged → Progressing=False ✓
- **Real upgrade (operator updates spec):** Generation bumps → Progressing=True ✓
- **Fresh install:** Generation=1, ObservedGeneration=0 → Progressing=True ✓

The Available condition (`IsAvailable` at line 231) independently monitors pod readiness and is unaffected.

## Test plan
- [x] `make` builds successfully
- [x] `make test-unit` all tests pass
- [x] `gofmt` / `go vet` clean
- [x] Unit tests for `checkDeploymentGenerationProgress`: ObservedGeneration equal, less than, greater than Generation, both zero, and initial rollout
- [x] Manual verification on live OCP 4.22 cluster (node drain stays Progressing=False, spec change triggers Progressing=True)
- [ ] CI e2e upgrade test `[Monitor:legacy-cvo-invariants][bz-Management Console] clusteroperator/console should stay Progressing=False while MCO is Progressing=True` should stop failing (~8% failure rate on Sippy currently)

Assisted-by: Claude Code (Opus 4.6)

## Summary by CodeRabbit

* **Bug Fixes**
  * More reliable rollout status: deployments are now marked "progressing" only when the cluster hasn’t observed the latest deployment generation, avoiding noisy status flips caused by fluctuating replica counts during external disruptions.

* **Tests**
  * Added tests covering observed-generation scenarios (equal, less-than, greater-than, and initial rollout) to validate consistent progression reporting.

### CODE REVIEW

File: pkg/console/operator/sync_v400.go [hunk 1]
```diff
@@ -211,13 +211,13 @@ func (co *consoleOperator) sync_v400(ctx context.Context, controllerContext fact
...
-		// "Operators should not report Progressing when they are reconciling (without action)
-		// a previously known state."
-		if !deploymentsub.IsAvailableAndUpdated(actualDeployment) {
-			return fmt.Errorf("working toward version %s, %v replicas available", version, actualDeployment.Status.AvailableReplicas)
+		// Only report Progressing when the deployment controller hasn't yet
+		// processed a spec change made by the operator. Do not use replica
+		// counts — they fluctuate during external disruptions (node reboots)
+		// that the operator did not cause. See https://issues.redhat.com/browse/OCPBUGS-64688
+		if actualDeployment.Status.ObservedGeneration < actualDeployment.Generation {
+			return fmt.Errorf("deployment generation %d not yet observed (current: %d)",
```
Comment (Leo6Leo) [inline pkg/console/operator/sync_v400.go]: The parenthetical says "current" but passes ObservedGeneration. Consider changing the wording to make it more accurate:
```suggestion
			return fmt.Errorf("deployment generation %d not yet observed (observed: %d)",
```

File: pkg/console/operator/sync_v400.go
Comment (Leo6Leo) [inline pkg/console/operator/sync_v400.go:1]: @sg00dwin Could you add some unit tests for the new behavior?

The old IsAvailableAndUpdated at least had its own unit tests in deployment_test.go. The new inline comparison ObservedGeneration < Generation has no test coverage. A table-driven test with cases like: - ObservedGeneration == Generation → no error (not progressing) - ObservedGeneration < Generation → returns error (progressing) - ObservedGeneration > Generation → no error (edge case, already processed)

would help prevent regressions.

**Comment by Leo6Leo** [comment]
/cc @Leo6Leo 

Spinning cluster to do QE currently

**Comment by Leo6Leo** [comment]
Manual Verification

Deployed the PR build to a live OCP 4.22 cluster (3 control-plane nodes, Azure).

Node drain (should stay Progressing=False): Drained a node hosting a console pod. Pod was evicted and rescheduled. Progressing stayed False throughout — lastTransitionTime never changed.

Spec change (should trigger Progressing=True): Changed logLevel to TraceAll, bumping deployment generation 3→4. Progressing correctly flipped to True ("deployment generation 4 not yet observed (current: 3)"), then resolved to False 2s later.

Both scenarios behave as expected.

Currently just waiting for the CI to past before I give the verify label.

File: pkg/console/operator/sync_v400.go
Comment (theobarberbany) [reply pkg/console/operator/sync_v400.go:1]: I'd also be considering adding an e2e to one of the upgrade jobs? If we can write one that isn't a giant pain.

Comment (sg00dwin) [reply pkg/console/operator/sync_v400.go:1]: **Unit tests:** Added table-driven unit tests for the ObservedGeneration-based Progressing check covering: equal, less than, greater than, both zero, and initial rollout scenarios.

**Regarding e2e test:** Good idea. There's already a story tracking this under the same epic: [CONSOLE-5188 (e2e testing automation)](https://redhat.atlassian.net/browse/CONSOLE-5188). It's currently unassigned and empty description. Added this suggestion for adding a console-owned e2e assertion to one of the upgrade jobs.

**Review by TheRealJon — APPROVED** [review]
/lgtm

**Review by jhadvig — APPROVED** [review]
/verified by CI
